bob-access/src/extractor.rs

```diff
@@ -0,0 +1,119 @@
+use crate::{credentials::{Credentials, CredentialsBuilder}, error::Error};
+use axum::extract::RequestParts;
+use http::{Extensions, HeaderMap, Request};
+use tonic::{transport::server::TcpConnectInfo};
+use super::{credentials_type, CredentialsType};
+
+pub trait Extractor {
+    fn headers(&self) -> Option<&HeaderMap>;
+    fn extensions(&self) -> Option<&Extensions>;
+}
+
+pub trait ExtractorExt {
+    fn extract(&self) -> Result<Credentials, Error>;
+    fn extract_basic(
+        &self,
+        _header_map: &HeaderMap,
+        _builder: CredentialsBuilder,
+    ) -> Result<Credentials, Error>;
+
+    fn extract_token(
+        &self,
+        _header_map: &HeaderMap,
+        _builder: CredentialsBuilder,
+    ) -> Result<Credentials, Error>;
+}
+
+fn prepare_builder<T: Extractor>(slf: &T) -> Result<(&HeaderMap, CredentialsBuilder), Error> {
+    let header_map = if let Some(header_map) = slf.headers() {
+        header_map
+    } else {
+        return Err(Error::CredentialsNotProvided("can't extract headers".into()));
+    };
+    let addr = slf
+        .extensions()
+        .and_then(|ext| ext.get::<TcpConnectInfo>()?.remote_addr());
+    let mut builder = Credentials::builder();
+    builder.with_address(addr);
+    Ok((header_map, builder))
+}
+
+impl<T: Extractor> ExtractorExt for T {
+    fn extract(&self) -> Result<Credentials, Error> {
+        let cred_tp = credentials_type();
+        match cred_tp {
+            CredentialsType::Stub => {
+                return Ok(Credentials::default());
+            },
+            CredentialsType::Basic => {
+                let (header_map, builder) = prepare_builder(self)?;
+                return Ok(self.extract_basic(header_map, builder)?);
+            },
+            CredentialsType::Token => {
+                let (header_map, builder) = prepare_builder(self)?;
+                return Ok(self.extract_token(header_map, builder)?);
+            },
+        }
+    }
+
+    fn extract_basic(
```

Comment:
I just noticed that `extract_basic` and `extract_token` are part of `ExtractorExt` trait and are actually available from outside. In this situation, these functions should be self-contained: they should not take `header_map` and `builder` as parameters, but perform all data extraction inside their body. Basically, that means that `prepare_builder` should be called within these functions body